driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/vault/ClientEncryptionImpl.java

```diff
@@ -183,6 +198,85 @@ public Publisher<RewrapManyDataKeyResult> rewrapManyDataKey(final Bson filter, f
                         }));
     }
 
+    @Override
+    public Publisher<BsonDocument> createEncryptedCollection(final MongoDatabase database, final String collectionName,
```

Comment:
@rozza, your knowledge of the Project Reactor and the reactive streams programming is being needed when reviewing this method.

The intended behavior of the implementation of this method can be peeked either in the synchronous implementation, or in the spec: [Create Encrypted Collection Helper](https://github.com/mongodb/specifications/blob/master/source/client-side-encryption/client-side-encryption.rst#create-encrypted-collection-helper).

The thing that made this implementation really difficult for me (beyond the lack of knowledge are):

- if either `createDataKey` or `createCollection` fails, the method must attach to the error signal the (possibly partially) updated `encryptedFields` that it would have produced in case of a success.

I achieved the needed behavior by capturing `dataKeyMayBeCreated` and `actualEncryptedFields` in the returned `Publisher`. As far as I understand, the `Publisher` is constructed such that there are no data races when updating the `actualEncryptedFields` document (updates are done to the unrelated `field` documents), and all reads of the updated `actualEncryptedFields` are ordered after (in the HB order) the updates because of the combination of `then` and `defer` methods.

Even if this implementation is correct (at the very least I doubt it is idiomatic), it can't work if there are more than one `Subscriber` (regardless of whether they exist concurrently or not), so had make the result a one-shot publisher.